security/security-api/custom-identity-store/README.md

```diff
@@ -0,0 +1,175 @@
+
+Defining credentials and the roles assigned to the users is done as shown below in `TestIdentityStore`:
+
+```java
+if (usernamePasswordCredential.compareTo("reza", "secret1")) {
+    return new CredentialValidationResult("reza", new HashSet<>(asList("foo", "bar")));
+}
+
+```
+
+With application, a custom `HttpAuthenticationMechanism` too is bundled which gets called during authentication process.
+
+```java
+@RequestScoped
+public class TestAuthenticationMechanism implements HttpAuthenticationMechanism {
+
+    @Inject
+    private IdentityStoreHandler identityStoreHandler;
+
+    @Override
+    public AuthenticationStatus validateRequest(HttpServletRequest request, HttpServletResponse response, HttpMessageContext httpMessageContext) throws AuthenticationException {
+
+        if (request.getParameter("name") != null && request.getParameter("password") != null) {
+
+            // Get the (caller) name and password from the request
+            // NOTE: This is for the smallest possible example only. In practice
+            // putting the password in a request query parameter is highly
+            // insecure
+            String name = request.getParameter("name");
+            Password password = new Password(request.getParameter("password"));
+
+            // Delegate the {credentials in -> identity data out} function to
+            // the Identity Store
+            // Password internally gets verified using the PasswordHash specified while defining DatabaseIdentityStore.
+            CredentialValidationResult result = identityStoreHandler.validate(
+                new UsernamePasswordCredential(name, password));
+
+            if (result.getStatus() == VALID) {
+                // Communicate the details of the authenticated user to the
+                // container. In many cases the underlying handler will just store the details
+                // and the container will actually handle the login after we return from
+                // this method.
+                return httpMessageContext.notifyContainerAboutLogin(
+                    result.getCallerPrincipal(), result.getCallerGroups());
+            } else {
+                return httpMessageContext.responseUnauthorized();
+            }
+        }
+
+        return httpMessageContext.doNothing();
+    }
+
+}
+```
+Following snippet from `TestAuthenticationMechanism` delegates the validation task to underlying `DefaultIdentityStoreHandler`, which in turn validates the credentials against `TestIdentityStore` as defined in this application.
+
+```java
+CredentialValidationResult result = identityStoreHandler.validate(
+    new UsernamePasswordCredential(name, password));
+```
+
+On successful authorization, container gets notified with caller details.
+
+```java
+return httpMessageContext.notifyContainerAboutLogin(
+    result.getCallerPrincipal(), result.getCallerGroups());
```

Comment:
Similar comment to above -- use `result.getCallerPrincipal().getName()`? It's probably reasonable to just use `result.getCallerPrincipal()` sometimes; we should show both methods in the various examples.